parallel-twitter-trending-hashtags/twitter_wallaroo_app.py

```diff
@@ -0,0 +1,173 @@
+import string
+import struct
+import time
+import wallaroo
+
+# Maintain list of TOP_X size for trending hashtags
+TOP_X = 10
+# Look for values over a TRENDING_OVER minute period
+TRENDING_OVER = 5
+
+
+def application_setup(args):
+    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
+    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
+
+    raw_hashtag_partitions = list(string.ascii_lowercase)
+    raw_hashtag_partitions.append("!")
+
+    ab = wallaroo.ApplicationBuilder("Trending Hashtags")
+    ab.new_pipeline("Tweets", wallaroo.TCPSourceConfig(in_host, in_port,
+                                                       decoder))
+    ab.to_parallel(find_hashtags)
+    ab.to_state_partition(count_hashtags, HashtagCounts, "raw hashtag counts",
+                          extract_hashtag_key, raw_hashtag_partitions)
+    ab.to_stateful(top_hashtags, TopTags, "top hashtags")
+    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, encoder))
+    return ab.build()
+
+
+@wallaroo.computation_multi(name="find hashtags")
+def find_hashtags(data):
+    punctuation = " !\"$%&\'()*+,-./:;<=>?@[\\]^_`{|}~"
+
+    hashtags = []
+
+    for line in data.split("\n"):
+        for word in line.split(" "):
+            clean_word = word.lower().strip(punctuation)
+            if len(clean_word) > 1 and clean_word[0] == "#":
+                hashtag = clean_word.strip('#')
+                if hashtag:
+                    hashtags.append(hashtag)
+
+    return hashtags
+
+
+@wallaroo.state_computation(name="count hashtags")
+def count_hashtags(hashtag, counts):
+    """
+    get the current top ten, if this changes
+    after updating, we'll send an updated list
+    to our top ten aggregator otherwise, skip
+    """
+    top_ten_if_changed = counts.increment(hashtag)
+    return (top_ten_if_changed, True)
+
+
+@wallaroo.state_computation(name="top tags")
+def top_hashtags(tags_and_counts, top_tags):
+    """
+    receive a dictionary of tags and counts,
+    add to our list of top tags and shrink to
+    only ten, if our new top ten is different than
+    our old, output the new list
+    """
+    top_ten_if_changed = top_tags.update(tags_and_counts)
+    return (top_ten_if_changed, top_ten_if_changed != None)
+
+
+class HashtagCounts(object):
+    def __init__(self):
+        self.__top_tags = {}
```

Comment:
Is there a reason this code uses double underscores instead of single underscores for members? Double underscores have a specific meaning in Python (https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-a-single-and-a-double-underscore-before-an-object-name) related to name mangling to avoid problems when creating subclasses. None of this code seems destined for subclassing, so I think folks might wonder why this is being done.

Reply:
That is not what I read. I read multiple things that say __X__ is a common way to indicate "this is intended to be private".

Reply:
Everything I read said that __X and __X__ are for marking things as "private". Which also appears to be the case based on that stackoverflow link you sent.

Also I believe __init__ is a magic python object constructor.

Reply:
Yeah the underscores around init are required. Using the double underscore makes me, as a Python programmer, think that something special might be happening, because as far as I know using a single underscore is sufficient to indicate to the client that they should not be using the member. The client should consider it private. When I see a double underscore I think that the author is conveying something more, that this member *really* needs to be hidden, so the name will be mangled. That makes me start wondering why this is being done. I start looking to see what's going on that necessitates this extra level of "private". And in this particular case there's nothing.

But maybe @nisanharamati has other thoughts.

Reply:
I agree with @aturley. These should be either one underscore, or no underscores (except `__init__` which is special)

Reply:
the "no underscores" option is because I'm not sure there's any benefit in having these demarcated as private. This isn't a module that's going to be imported and used by someone else.
The same applies to the other classes in this file.